File: app/src/main/java/com/offlinetranslator/app/feature/backup/BackupCodec.kt
package com.offlinetranslator.app.feature.backup

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

/**
 * 备份编解码 + 合并去重纯函数。无 Android DB/IO/UI 依赖，
 * 仅用 [org.json] 处理跨平台 JSON v1。
 */
object BackupCodec {

    // ---------- encode ----------

    fun encode(backup: Backup): String {
        val root = JSONObject()
        root.put("format", backup.format)
        root.put("version", backup.version)
        root.put("exportedAt", backup.exportedAt)
        root.put("platform", backup.platform)

        val books = JSONArray()
        for (book in backup.wordBooks) books.put(encodeBook(book))
        root.put("wordBooks", books)

        val translations = JSONArray()
        for (t in backup.translations) translations.put(encodeTranslation(t))
        root.put("translations", translations)

        val chats = JSONArray()
        for (c in backup.chats) chats.put(encodeChat(c))
        root.put("chats", chats)

        backup.config?.let { root.put("config", encodeConfig(it)) }

        return root.toString(2)
    }

    private fun encodeSrs(srs: BackupSrs): JSONObject = JSONObject().apply {
        put("box", srs.box)
        put("dueAt", srs.dueAt)
        put("missCount", srs.missCount)
        put("lastReviewedAt", srs.lastReviewedAt)
    }

    private fun encodeEntry(entry: BackupEntry): JSONObject = JSONObject().apply {
        put("english", entry.english)
        put("chinese", entry.chinese)
        put("note", entry.note)
        put("createdAt", entry.createdAt)
        entry.srs?.let { put("srs", encodeSrs(it)) }
    }

    private fun encodeBook(book: BackupBook): JSONObject = JSONObject().apply {
        put("name", book.name)
        put("purpose", book.purpose)
        put("dailyGoal", book.dailyGoal)
        put("createdAt", book.createdAt)
        val entries = JSONArray()
        for (e in book.entries) entries.put(encodeEntry(e))
        put("entries", entries)
    }

    private fun encodeTranslation(t: BackupTranslation): JSONObject = JSONObject().apply {
        put("sourceText", t.sourceText)
        put("translatedText", t.translatedText)
        put("sourceLang", t.sourceLang)
        put("targetLang", t.targetLang)
        put("createdAt", t.createdAt)
        put("starred", t.starred)
        t.srs?.let { put("srs", encodeSrs(it)) }
    }

    private fun encodeMessage(m: BackupMessage): JSONObject = JSONObject().apply {
        put("role", m.role)
        put("content", m.content)
        put("createdAt", m.createdAt)
    }

    private fun encodeChat(c: BackupChat): JSONObject = JSONObject().apply {
        put("id", c.id)
        put("title", c.title)
        put("updatedAt", c.updatedAt)
        put("modelId", c.modelId)
        c.summary?.let { put("summary", it) }
        put("summarizedCount", c.summarizedCount)
        val messages = JSONArray()
        for (m in c.messages) messages.put(encodeMessage(m))
        put("messages", messages)
    }

    private fun encodeConfig(c: BackupConfig): JSONObject = JSONObject().apply {
        put("reminderEnabled", c.reminderEnabled)
        put("reminderHour", c.reminderHour)
        put("reminderMinute", c.reminderMinute)
        c.uiLanguage?.let { put("uiLanguage", it) }
        c.activeModelId?.let { put("activeModelId", it) }
        put("streakLastDay", c.streakLastDay)
        put("streakCurrent", c.streakCurrent)
        put("streakLongest", c.streakLongest)
    }

    // ---------- decode ----------

    fun decode(json: String): Result<Backup> {
        val root = try {
            JSONObject(json)
        } catch (e: JSONException) {
            return Result.failure(IllegalArgumentException("非法 JSON: ${e.message}"))
        }

        val format = root.optString("format", "")
        if (format != BACKUP_FORMAT) {
            return Result.failure(IllegalArgumentException("格式不匹配: format=$format"))
        }
        val version = root.optInt("version", 0)
        if (version !in 1..BACKUP_VERSION) {
            return Result.failure(IllegalArgumentException("不支持的版本: version=$version"))
        }

        return try {
            val backup = Backup(
                format = format,
                version = version,
                exportedAt = root.optLong("exportedAt", 0L),
                platform = root.optString("platform", ""),
                wordBooks = decodeBooks(root.optJSONArray("wordBooks")),
                translations = decodeTranslations(root.optJSONArray("translations")),
                chats = decodeChats(root.optJSONArray("chats")),
                config = root.optJSONObject("config")?.let { decodeConfig(it) },
            )
            Result.success(backup)
        } catch (e: JSONException) {
            Result.failure(IllegalArgumentException("解析失败: ${e.message}"))
        }
    }

    private fun decodeSrs(obj: JSONObject?): BackupSrs? {
        if (obj == null) return null
        return BackupSrs(
            box = obj.optInt("box", 0),
            dueAt = obj.optLong("dueAt", 0L),
            missCount = obj.optInt("missCount", 0),
            lastReviewedAt = obj.optLong("lastReviewedAt", 0L),
        )
    }

    private fun decodeEntry(obj: JSONObject): BackupEntry = BackupEntry(
        english = obj.optString("english", ""),
        chinese = obj.optString("chinese", ""),
        note = obj.optString("note", ""),
        createdAt = obj.optLong("createdAt", 0L),
        srs = decodeSrs(obj.optJSONObject("srs")),
    )

    private fun decodeBooks(arr: JSONArray?): List<BackupBook> {
        if (arr == null) return emptyList()
        val out = ArrayList<BackupBook>(arr.length())
        for (i in 0 until arr.length()) {
            val obj = arr.getJSONObject(i)
            val entriesArr = obj.optJSONArray("entries")
            val entries = if (entriesArr == null) emptyList() else
                (0 until entriesArr.length()).map { decodeEntry(entriesArr.getJSONObject(it)) }
            out.add(
                BackupBook(
                    name = obj.optString("name", ""),
                    purpose = obj.optString("purpose", ""),
                    dailyGoal = obj.optInt("dailyGoal", 0),
                    createdAt = obj.optLong("createdAt", 0L),
                    entries = entries,
                )
            )
        }
        return out
    }

    private fun decodeTranslations(arr: JSONArray?): List<BackupTranslation> {
        if (arr == null) return emptyList()
        val out = ArrayList<BackupTranslation>(arr.length())
        for (i in 0 until arr.length()) {
            val obj = arr.getJSONObject(i)
            out.add(
                BackupTranslation(
                    sourceText = obj.optString("sourceText", ""),
                    translatedText = obj.optString("translatedText", ""),
                    sourceLang = obj.optString("sourceLang", ""),
                    targetLang = obj.optString("targetLang", ""),
                    createdAt = obj.optLong("createdAt", 0L),
                    starred = obj.optBoolean("starred", false),
                    srs = decodeSrs(obj.optJSONObject("srs")),
                )
            )
        }
        return out
    }

    private fun decodeChats(arr: JSONArray?): List<BackupChat> {
        if (arr == null) return emptyList()
        val out = ArrayList<BackupChat>(arr.length())
        for (i in 0 until arr.length()) {
            val obj = arr.getJSONObject(i)
            val msgArr = obj.optJSONArray("messages")
            val messages = if (msgArr == null) emptyList() else
                (0 until msgArr.length()).map {
                    val m = msgArr.getJSONObject(it)
                    BackupMessage(
                        role = m.optString("role", ""),
                        content = m.optString("content", ""),
                        createdAt = m.optLong("createdAt", 0L),
                    )
                }
            out.add(
                BackupChat(
                    id = obj.optString("id", ""),
                    title = obj.optString("title", ""),
                    updatedAt = obj.optLong("updatedAt", 0L),
                    modelId = obj.optString("modelId", ""),
                    summary = if (obj.has("summary") && !obj.isNull("summary")) obj.optString("summary") else null,
                    summarizedCount = obj.optInt("summarizedCount", 0),
                    messages = messages,
                )
            )
        }
        return out
    }

    private fun decodeConfig(obj: JSONObject): BackupConfig = BackupConfig(
        reminderEnabled = obj.optBoolean("reminderEnabled", false),
        reminderHour = obj.optInt("reminderHour", 0),
        reminderMinute = obj.optInt("reminderMinute", 0),
        uiLanguage = if (obj.has("uiLanguage") && !obj.isNull("uiLanguage")) obj.optString("uiLanguage") else null,
        activeModelId = if (obj.has("activeModelId") && !obj.isNull("activeModelId")) obj.optString("activeModelId") else null,
        streakLastDay = obj.optLong("streakLastDay", 0L),
        streakCurrent = obj.optInt("streakCurrent", 0),
        streakLongest = obj.optInt("streakLongest", 0),
    )

    // ---------- merge ----------

    fun merge(current: Backup, incoming: Backup): MergePlan {
        // Books by trimmed name.
        val currentBooksByKey = current.wordBooks.associateBy { it.name.trim() }
        val newBooks = ArrayList<BackupBook>()
        val entriesToExistingBook = LinkedHashMap<String, List<BackupEntry>>()

        for (book in incoming.wordBooks) {
            val key = book.name.trim()
            val existing = currentBooksByKey[key]
            if (existing == null) {
                newBooks.add(book)
            } else {
                val existingEntryKeys = existing.entries
                    .map { it.english.trim().lowercase() }
                    .toHashSet()
                val toAdd = book.entries.filter {
                    it.english.trim().lowercase() !in existingEntryKeys
                }
                if (toAdd.isNotEmpty()) {
                    entriesToExistingBook[book.name] = toAdd
                }
            }
        }

        // Translations by (sourceText, sourceLang, targetLang).
        val currentTransKeys = current.translations
            .map { transKey(it.sourceText, it.sourceLang, it.targetLang) }
            .toHashSet()
        val newTranslations = incoming.translations.filter {
            transKey(it.sourceText, it.sourceLang, it.targetLang) !in currentTransKeys
        }

        // Chats by id.
        val currentChatIds = current.chats.map { it.id }.toHashSet()
        val newChats = incoming.chats.filter { it.id !in currentChatIds }

        // Config / streak resolution.
        val resolvedConfig = incoming.config?.let { inc ->
            val cur = current.config
            val resolvedLongest = maxOf(cur?.streakLongest ?: 0, inc.streakLongest)
            val (resolvedLastDay, resolvedCurrent) =
                if (cur == null || inc.streakLastDay > cur.streakLastDay) {
                    inc.streakLastDay to inc.streakCurrent
                } else {
                    cur.streakLastDay to cur.streakCurrent
                }
            inc.copy(
                streakLastDay = resolvedLastDay,
                streakCurrent = resolvedCurrent,
                streakLongest = resolvedLongest,
            )
        }

        val addedEntries = newBooks.sumOf { it.entries.size } +
            entriesToExistingBook.values.sumOf { it.size }

        return MergePlan(
            newBooks = newBooks,
            entriesToExistingBook = entriesToExistingBook,
            newTranslations = newTranslations,
            newChats = newChats,
            resolvedConfig = resolvedConfig,
            addedBooks = newBooks.size,
            addedEntries = addedEntries,
            addedTranslations = newTranslations.size,
            addedChats = newChats.size,
        )
    }

    private fun transKey(sourceText: String, sourceLang: String, targetLang: String): String =
        "$sourceText $sourceLang $targetLang"
}

File: ios/Yiren/Features/Backup/BackupCodec.swift
import Foundation

// 跨平台备份数据模型（JSON v1）。字段名与嵌套结构即为权威格式，
// 与安卓端 com.offlinetranslator.app.feature.backup 逐字段镜像，切勿擅自改名。

struct BackupSrs: Codable, Equatable {
    let box: Int
    let dueAt: Int64
    let missCount: Int
    let lastReviewedAt: Int64
}

struct BackupEntry: Codable, Equatable {
    let english: String
    let chinese: String
    let note: String
    let createdAt: Int64
    let srs: BackupSrs?
}

struct BackupBook: Codable, Equatable {
    let name: String
    let purpose: String
    let dailyGoal: Int
    let createdAt: Int64
    let entries: [BackupEntry]

    enum CodingKeys: String, CodingKey {
        case name, purpose, dailyGoal, createdAt, entries
    }

    init(name: String, purpose: String, dailyGoal: Int, createdAt: Int64, entries: [BackupEntry]) {
        self.name = name
        self.purpose = purpose
        self.dailyGoal = dailyGoal
        self.createdAt = createdAt
        self.entries = entries
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        name = try c.decodeIfPresent(String.self, forKey: .name) ?? ""
        purpose = try c.decodeIfPresent(String.self, forKey: .purpose) ?? ""
        dailyGoal = try c.decodeIfPresent(Int.self, forKey: .dailyGoal) ?? 0
        createdAt = try c.decodeIfPresent(Int64.self, forKey: .createdAt) ?? 0
        entries = try c.decodeIfPresent([BackupEntry].self, forKey: .entries) ?? []
    }
}

struct BackupTranslation: Codable, Equatable {
    let sourceText: String
    let translatedText: String
    let sourceLang: String
    let targetLang: String
    let createdAt: Int64
    let starred: Bool
    let srs: BackupSrs?
}

struct BackupMessage: Codable, Equatable {
    let role: String
    let content: String
    let createdAt: Int64
}

struct BackupChat: Codable, Equatable {
    let id: String
    let title: String
    let updatedAt: Int64
    let modelId: String
    let summary: String?
    let summarizedCount: Int
    let messages: [BackupMessage]

    enum CodingKeys: String, CodingKey {
        case id, title, updatedAt, modelId, summary, summarizedCount, messages
    }

    init(id: String, title: String, updatedAt: Int64, modelId: String, summary: String?, summarizedCount: Int, messages: [BackupMessage]) {
        self.id = id
        self.title = title
        self.updatedAt = updatedAt
        self.modelId = modelId
        self.summary = summary
        self.summarizedCount = summarizedCount
        self.messages = messages
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(String.self, forKey: .id) ?? ""
        title = try c.decodeIfPresent(String.self, forKey: .title) ?? ""
        updatedAt = try c.decodeIfPresent(Int64.self, forKey: .updatedAt) ?? 0
        modelId = try c.decodeIfPresent(String.self, forKey: .modelId) ?? ""
        summary = try c.decodeIfPresent(String.self, forKey: .summary)
        summarizedCount = try c.decodeIfPresent(Int.self, forKey: .summarizedCount) ?? 0
        messages = try c.decodeIfPresent([BackupMessage].self, forKey: .messages) ?? []
    }
}

struct BackupConfig: Codable, Equatable {
    let reminderEnabled: Bool
    let reminderHour: Int
    let reminderMinute: Int
    let uiLanguage: String?
    let activeModelId: String?
    let streakLastDay: Int64
    let streakCurrent: Int
    let streakLongest: Int
}

struct Backup: Codable, Equatable {
    let format: String
    let version: Int
    let exportedAt: Int64
    let platform: String
    let wordBooks: [BackupBook]
    let translations: [BackupTranslation]
    let chats: [BackupChat]
    let config: BackupConfig?

    enum CodingKeys: String, CodingKey {
        case format, version, exportedAt, platform, wordBooks, translations, chats, config
    }

    init(format: String, version: Int, exportedAt: Int64, platform: String,
         wordBooks: [BackupBook], translations: [BackupTranslation], chats: [BackupChat],
         config: BackupConfig?) {
        self.format = format
        self.version = version
        self.exportedAt = exportedAt
        self.platform = platform
        self.wordBooks = wordBooks
        self.translations = translations
        self.chats = chats
        self.config = config
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        format = try c.decodeIfPresent(String.self, forKey: .format) ?? ""
        version = try c.decodeIfPresent(Int.self, forKey: .version) ?? 0
        exportedAt = try c.decodeIfPresent(Int64.self, forKey: .exportedAt) ?? 0
        platform = try c.decodeIfPresent(String.self, forKey: .platform) ?? ""
        wordBooks = try c.decodeIfPresent([BackupBook].self, forKey: .wordBooks) ?? []
        translations = try c.decodeIfPresent([BackupTranslation].self, forKey: .translations) ?? []
        chats = try c.decodeIfPresent([BackupChat].self, forKey: .chats) ?? []
        config = try c.decodeIfPresent(BackupConfig.self, forKey: .config)
    }
}

struct MergePlan: Equatable {
    let newBooks: [BackupBook]
    let entriesToExistingBook: [String: [BackupEntry]]
    let newTranslations: [BackupTranslation]
    let newChats: [BackupChat]
    let resolvedConfig: BackupConfig?
    let addedBooks: Int
    let addedEntries: Int
    let addedTranslations: Int
    let addedChats: Int
}

let backupFormat = "yiren-backup"
let backupVersion = 1

struct BackupError: LocalizedError, Equatable {
    let message: String
    var errorDescription: String? { message }
}

enum BackupCodec {

    // ---------- encode ----------

    static func encode(_ backup: Backup) -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        // nil 可选字段（srs/summary/uiLanguage/activeModelId/config）由合成 Codable 自动省略，不会写成 null。
        return (try? encoder.encode(backup)) ?? Data()
    }

    // ---------- decode ----------

    static func decode(_ data: Data) -> Result<Backup, Error> {
        let decoder = JSONDecoder()
        let backup: Backup
        do {
            backup = try decoder.decode(Backup.self, from: data)
        } catch {
            return .failure(BackupError(message: "无法识别的备份文件"))
        }
        guard backup.format == backupFormat, backup.version == backupVersion else {
            return .failure(BackupError(message: "无法识别的备份文件"))
        }
        return .success(backup)
    }

    // ---------- merge ----------

    static func merge(current: Backup, incoming: Backup) -> MergePlan {
        // Books by trimmed name.
        var currentBooksByKey: [String: BackupBook] = [:]
        for book in current.wordBooks {
            currentBooksByKey[book.name.trimmingCharacters(in: .whitespaces)] = book
        }

        var newBooks: [BackupBook] = []
        var entriesToExistingBook: [String: [BackupEntry]] = [:]

        for book in incoming.wordBooks {
            let key = book.name.trimmingCharacters(in: .whitespaces)
            guard let existing = currentBooksByKey[key] else {
                newBooks.append(book)
                continue
            }
            let existingEntryKeys = Set(existing.entries.map {
                $0.english.trimmingCharacters(in: .whitespaces).lowercased()
            })
            let toAdd = book.entries.filter {
                !existingEntryKeys.contains($0.english.trimmingCharacters(in: .whitespaces).lowercased())
            }
            if !toAdd.isEmpty {
                entriesToExistingBook[book.name] = toAdd
            }
        }

        // Translations by (sourceText, sourceLang, targetLang).
        let currentTransKeys = Set(current.translations.map {
            transKey($0.sourceText, $0.sourceLang, $0.targetLang)
        })
        let newTranslations = incoming.translations.filter {
            !currentTransKeys.contains(transKey($0.sourceText, $0.sourceLang, $0.targetLang))
        }

        // Chats by id.
        let currentChatIds = Set(current.chats.map { $0.id })
        let newChats = incoming.chats.filter { !currentChatIds.contains($0.id) }

        // Config / streak resolution.
        let resolvedConfig: BackupConfig?
        if let inc = incoming.config {
            let cur = current.config
            let resolvedLongest = max(cur?.streakLongest ?? 0, inc.streakLongest)
            let resolvedLastDay: Int64
            let resolvedCurrent: Int
            if cur == nil || inc.streakLastDay > cur!.streakLastDay {
                resolvedLastDay = inc.streakLastDay
                resolvedCurrent = inc.streakCurrent
            } else {
                resolvedLastDay = cur!.streakLastDay
                resolvedCurrent = cur!.streakCurrent
            }
            resolvedConfig = BackupConfig(
                reminderEnabled: inc.reminderEnabled,
                reminderHour: inc.reminderHour,
                reminderMinute: inc.reminderMinute,
                uiLanguage: inc.uiLanguage,
                activeModelId: inc.activeModelId,
                streakLastDay: resolvedLastDay,
                streakCurrent: resolvedCurrent,
                streakLongest: resolvedLongest
            )
        } else {
            resolvedConfig = nil
        }

        let addedEntries = newBooks.reduce(0) { $0 + $1.entries.count } +
            entriesToExistingBook.values.reduce(0) { $0 + $1.count }

        return MergePlan(
            newBooks: newBooks,
            entriesToExistingBook: entriesToExistingBook,
            newTranslations: newTranslations,
            newChats: newChats,
            resolvedConfig: resolvedConfig,
            addedBooks: newBooks.count,
            addedEntries: addedEntries,
            addedTranslations: newTranslations.count,
            addedChats: newChats.count
        )
    }

    private static func transKey(_ sourceText: String, _ sourceLang: String, _ targetLang: String) -> String {
        "\(sourceText) \(sourceLang) \(targetLang)"
    }
}
